app/atomic_svc.py

```diff
@@ -0,0 +1,234 @@
+from app.utility.base_service import BaseService
+
+PLATFORMS = dict(windows='windows', macos='darwin', linux='linux')
+EXECUTORS = dict(command_prompt='cmd', sh='sh', powershell='psh', bash='sh')
+RE_VARIABLE = re.compile('(#{(.*?)})', re.DOTALL)
+PREFIX_HASH_LEN = 6
+
+
+class AtomicService(BaseService):
+
+    def __init__(self, services):
+        self.log = self.add_service('atomic_svc', self)
+
+        # Atomic Red Team attacks don't come with the corresponding tactic (phase name)
+        # so we need to create a match between techniques and tactics.
+        # This variable is filled by self._populate_dict_techniques_tactics()
+        self.technique_to_tactics = defaultdict(list)
+
+        self.atomic_dir = os.path.join('plugins', 'atomic')
+        self.repo_dir = os.path.join(self.atomic_dir, 'atomic-red-team')
+        self.data_dir = os.path.join(self.atomic_dir, 'data')
+        self.payloads_dir = os.path.join(self.atomic_dir, 'payloads')
+
+    async def clone_atomic_red_team_repo(self, repo_url=None):
+        """
+        Clone the Atomic Red Team repository. You can use a specific url via
+        the `repo_url` parameter (eg. if you want to use a fork).
+        """
+        if not repo_url:
+            repo_url = 'https://github.com/redcanaryco/atomic-red-team.git'
+
+        if not os.path.exists(self.repo_dir) or not os.listdir(self.repo_dir):
+            check_call(['git', 'clone', '--depth', '1', repo_url, self.repo_dir], stdout=DEVNULL, stderr=STDOUT)
+
+    async def populate_data_directory(self, path_yaml=None):
+        """
+        Populate the 'data' directory with the Atomic Red Team abilities.
+        These data will be usable by caldera after importation.
+        You can specify where the yaml files to import are located with the `path_yaml` parameter.
+        By default, read the yaml files in the atomics/ directory inside the Atomic Red Team repository.
+        """
+        if not self.technique_to_tactics:
+            await self._populate_dict_techniques_tactics()
+
+        if not path_yaml:
+            path_yaml = os.path.join(self.repo_dir, 'atomics', '**', 'T*.yaml')
+
+        at_total = 0
+        at_ingested = 0
+        errors = 0
+        for filename in glob.iglob(path_yaml):
+            for entries in BaseWorld.strip_yml(filename):
+                for test in entries['atomic_tests']:
+                    at_total += 1
+                    try:
+                        if await self._save_ability(entries, test):
+                            at_ingested += 1
+                    except Exception as e:
+                        self.log.debug('ERROR:', filename, e)
+                        errors += 1
+
+        errors_output = f' and ran into {errors} errors' if errors else ''
+        self.log.debug(f'Ingested {at_ingested} abilities (out of {at_total}) from Atomic plugin{errors_output}')
+
+    """ PRIVATE """
+
+    def _gen_single_match_tactic_technique(self, mitre_json):
```

Comment:
this function appears to be static-- let's label with @staticmethod

Reply:
fixed in 0948d05